### ISSUE TITLE
[QST] WIP int8 quantized impl but can't get transposed LDSM to work with current layout  

### ISSUE THREAD

**Issue by carlguo866** [Dao-AILab/flash-attention#1041]
I wrote a minimal int8 quantized Flash Attention implementation that sees a 31% speedup with reasonable accuracy. One big issue hindering its performance is that it can't utilize LDSM for the transposed V matrix and falls back on `DefaultCopy`. While the Q and K swizzled layout works rather naturally with a CopyAtom of `SM75_U32x2_LDSM_N`, I'm not sure how to construct a layout for V to match the CopyAtom `SM75_U16x4_LDSM_T`. The atom will require two elements per row and then going down through the columns (see below), and not sure we can construct a swizzled layout out of it either. Much appreciated if you have any insights!
![Screenshot 2024-07-10 at 11 57 57 AM](https://github.com/Dao-AILab/flash-attention/assets/30856909/4f1e83f1-e413-4a08-a523-bc7322f1baf9)

**Comment by tridao** [comment]
Right LDSM won't work for V if the data is 8bit. 
We might have some way to address this soon.

**Comment by carlguo866** [comment]
Could you provide a bit more detail on how we can approach this? Any idea is welcome, can be half-baked and I can explore more on my own. Happy to make a PR once my stuff gets more polished as well.

**Comment by tridao** [comment]
soon

**Comment by tridao** [comment]
you can use LDSM.T and byte-permute, then LDSM, as a way to transpose V
we'll release that code soon
idk if it works well without warp specialization

**Comment by carlguo866** [comment]
Looking forward to reading that code! I'm not sure exactly what u mean by doing 2 LDSMs? I thought we can only do one LDSM to load from smem to registers, and theres a barrier between the two.

**Comment by tridao** [comment]
Sorry i mean LDSM.T, byte permute, then store using STSM.

**Comment by carlguo866** [comment]
Got it - sadly STSM is a Hopper only instruction as well, but now I think about it, LDSM, byte-permute, and then STSM.T make more sense? Using LDSM.T would circle back to my original question, which is that the current swizzled layout doesn't allow LDSM.T in the first place because the smallest LDSM unit is 16bit and the instruction needs two elements per row. Register layout is much easier to modify (byte-permute) in comparison. I guess I can try changing the smem layout to fit what LDSM.T wants, but it doesn't seem clean and would cause more bank conflicts than swizzled layout

**Comment by tridao** [comment]
I see, I forgot that STSM is Hopper only.
The other option is to transpose V in a separate kernel, or fused it with a preceding kernel (e.g. gemm).

**Comment by carlguo866** [comment]
Yeah, makes sense - i dont think theres a preceding kernel using the same V and am worried a separate kernel would cause too much performance. I'll explore the different options and lyk if i have more updates

**Comment by carlguo866** [comment]
tentatively I found a solution to be using normal LSTM to load from smem to registers and then use `movmatrix` to transpose on registers

**Comment by itsliupeng** [comment]
> tentatively I found a solution to be using normal LSTM to load from smem to registers and then use `movmatrix` to transpose on registers

FYI.  FA3 has a solution using lamatrix and stmatrix instructions to implement in-kernel transpose for fp8 matrix.  https://github.com/Dao-AILab/flash-attention/blob/main/hopper/mainloop_fwd_sm90_tma_gmma_ws.hpp#L25

**Comment by carlguo866** [comment]
thx for the note - as Tri and I discussed above, stmatrix is only supported on Hopper gpus while I don't work with them
